src/internal/connector/onedrive/collections.go

```diff
@@ -502,277 +606,151 @@ func (c *Collections) UpdateCollections(
 	invalidPrevDelta bool,
 ) error {
 	for _, item := range items {
-		var (
-			prevPath           path.Path
-			prevCollectionPath path.Path
-			ok                 bool
-		)
+		itemID := ptr.Val(item.GetId())
+		isFolder := item.GetFolder() != nil || item.GetPackage() != nil
 
-		if item.GetRoot() != nil {
-			rootPath, err := GetCanonicalPath(
-				fmt.Sprintf(rootDrivePattern, driveID),
-				c.tenant,
-				c.resourceOwner,
-				c.source,
-			)
-			if err != nil {
-				return err
+		// Deleted file or folder.
+		if item.GetDeleted() != nil {
+			if err := c.handleDelete(
+				itemID,
+				driveID,
+				oldPaths,
+				newPaths,
+				isFolder,
+				excluded,
+			); err != nil {
+				return clues.Wrap(err, "marking item deleted")
 			}
 
-			// Add root path so as to have root folder information in
-			// path metadata. Root id -> path map is necessary in
-			// following delta incremental backups.
-			updatePath(newPaths, *item.GetId(), rootPath.String())
-
 			continue
 		}
 
-		if item.GetParentReference() == nil ||
-			item.GetParentReference().GetId() == nil ||
-			(item.GetDeleted() == nil && item.GetParentReference().GetPath() == nil) {
-			err := clues.New("no parent reference").With("item_id", *item.GetId())
-			if item.GetName() != nil {
-				err = err.With("item_name", *item.GetName())
-			}
-
-			return err
-		}
-
-		// Create a collection for the parent of this item
-		collectionID := *item.GetParentReference().GetId()
-
-		var collectionPathStr string
-		if item.GetDeleted() == nil {
-			collectionPathStr = *item.GetParentReference().GetPath()
-		} else {
-			collectionPathStr, ok = oldPaths[*item.GetParentReference().GetId()]
-			if !ok {
-				// This collection was created and destroyed in
-				// between the current and previous invocation
-				continue
-			}
-		}
-
-		collectionPath, err := GetCanonicalPath(
-			collectionPathStr,
-			c.tenant,
-			c.resourceOwner,
-			c.source,
-		)
+		collectionPath, err := c.getCollectionPath(driveID, item)
 		if err != nil {
 			return err
 		}
 
-		var (
-			itemPath path.Path
-			isFolder = item.GetFolder() != nil || item.GetPackage() != nil
-		)
-
-		if item.GetDeleted() == nil {
-			name := ptr.Val(item.GetName())
-			if len(name) == 0 {
-				return clues.New("non-deleted item with empty name").With("item_id", name)
-			}
-
-			itemPath, err = collectionPath.Append(name, !isFolder)
-			if err != nil {
-				return err
-			}
-		}
-
 		// Skip items that don't match the folder selectors we were given.
-		if shouldSkipDrive(ctx, itemPath, c.matcher, driveName) &&
-			shouldSkipDrive(ctx, collectionPath, c.matcher, driveName) {
+		if shouldSkipDrive(ctx, collectionPath, c.matcher, driveName) {
```

Comment:
Should `shouldSkipDrive` check happen before we delete item(call `handleDelete`)? We could add a test for this.

Reply:
I'm not sure that would give the behavior we want either :/ Unfortunately when items are deleted they have their parent reference set to the root folder. This is why we need the exclude list instead of just adding a deleted item entry in the collection for the item. Since they all end up in the root folder and we don't have information about paths from the previous backup the `shouldSkipDrive` check isn't particularly useful

I guess on the bright side, `shouldSkipDrive` really only comes into play for CI testing. Since the code was originally written we've moved away from allowing CLI users to backup individual folders/items and instead require backing up the entire service. This makes me a little less worried about edge cases I suppose